File: roms/tank_force.py
#
#	Tank Force
#

from base64 import b64encode
from io import BytesIO
from math import ceil
from sys import argv
from zipfile import ZipFile
from PIL import Image

with ZipFile(argv[1]) as z:
    snd = z.read('tf1_snd0.bin')
    prg = z.read('tf1_prg0.bin') + z.read('tf1_prg1.bin') + z.read('tankfrcej/tf1_prg7.bin')
    mcu = z.read('cus64-64a1.mcu')
    voi = z.read('tf1_voi0.bin') + z.read('tf1_voi1.bin')
    chr8 = z.read('tf1_chr8.bin')
    chr = z.read('tf1_chr0.bin') + z.read('tf1_chr1.bin') + z.read('tf1_chr2.bin') + z.read('tf1_chr3.bin') + z.read('tf1_chr4.bin') + z.read('tf1_chr5.bin')
    obj = z.read('tf1_obj0.bin') + z.read('tf1_obj1.bin')

rom = snd + prg + mcu + voi + chr8 + chr + obj

def pngstring(a):
    w = 1024
    img = Image.new('P', (w, ceil(len(a) / w)))
    img.putpalette(sum([[i, 0, 0] for i in range(256)], []))
    img.putdata(a)
    buf = BytesIO()
    img.save(buf, 'PNG')
    return b64encode(buf.getvalue())

with open(argv[2], 'wb') as f:
    f.write(b'export const ROM = \'data:image/png;base64,' + pngstring(rom) + b'\';\n')

File: roms/tank_force.js
/*
 *	Tank Force
 */

const fs = require('fs');
const {createCanvas, createImageData} = require('canvas');
const {Zlib} = require('zlibjs/bin/unzip.min');
Zlib.Unzip.prototype.read = Zlib.Unzip.prototype.decompress;

const cat = (...args) => {
	const array = new Uint8Array(args.reduce((a, b) => a + b.length, 0));
	return args.reduce((a, b) => (array.set(b, a), a + b.length), 0), array;
};

const z = new Zlib.Unzip(fs.readFileSync(process.argv[2]));
const snd = z.read('tf1_snd0.bin');
const prg = cat(z.read('tf1_prg0.bin'), z.read('tf1_prg1.bin'), z.read('tankfrcej/tf1_prg7.bin'));
const mcu = z.read('cus64-64a1.mcu');
const voi = cat(z.read('tf1_voi0.bin'), z.read('tf1_voi1.bin'));
const chr8 = z.read('tf1_chr8.bin');
const chr = cat(z.read('tf1_chr0.bin'), z.read('tf1_chr1.bin'), z.read('tf1_chr2.bin'), z.read('tf1_chr3.bin'), z.read('tf1_chr4.bin'), z.read('tf1_chr5.bin'));
const obj = cat(z.read('tf1_obj0.bin'), z.read('tf1_obj1.bin'));

const rom = cat(snd, prg, mcu, voi, chr8, chr, obj);

const pngString = array => {
	const w = 1024, h = Math.ceil(array.byteLength / w), canvas = createCanvas(w, h);
	const palette = new Uint8ClampedArray(new Uint32Array(256).map((e, i) => i | 0xff000000).buffer);
	canvas.getContext('2d', {pixelFormat: 'A8'}).putImageData(createImageData(new Uint8ClampedArray(array.buffer), w, h), 0, 0);
	return Buffer.from(canvas.toBuffer('image/png', {compressionLevel: 9, filters: canvas.PNG_FILTER_NONE, palette})).toString('base64');
};

fs.writeFileSync(process.argv[3], `export const ROM = 'data:image/png;base64,${pngString(rom)}';\n`);
